File: Trie/Implement Trie II/Implement_Trie_II.cpp
// Problem Name: Implement II
// Problem Link: https://www.naukri.com/code360/problems/implement-trie_1387095

#include <bits/stdc++.h> 

class Node {
public:
    Node* links[26];
    int cntEndsWith = 0;
    int cntPrefix = 0;

    bool containsKey(char ch) {
        return links[ch-'a'] != NULL;
    }

    Node* get(char ch) {
        return links[ch-'a']; 
    }

    void put(char ch, Node* node) {
        links[ch-'a'] = node;
    }

    void incPrefixCount() {
        cntPrefix++;
    }

    void decPrefixCount() {
        cntPrefix--;
    }

    void incEndsWith() {
        cntEndsWith++;
    }

    void decEndsWith() {
        cntEndsWith--;
    }

    int getPrefixCount() {
        return cntPrefix;
    }

    int getEndsWithCount() {
        return cntEndsWith;
    }
};

class Trie{
    private:

    Node* root;

    public:

    Trie(){
        root = new Node();
    }

    void insert(string &word){
        Node* node = root;
        for(int i = 0;i<word.size();i++) {
            if(!node->containsKey(word[i])) {
                node->put(word[i], new Node());
            }
            node = node->get(word[i]);
            node->incPrefixCount();
        }
        node->incEndsWith();
    }

    int countWordsEqualTo(string &word){
        Node* node = root;
        for(int i = 0;i<word.size();i++) {
            if(!node->containsKey(word[i])){
                return 0;
            }
            node = node->get(word[i]);
        }
        return node->getEndsWithCount();
    }

    int countWordsStartingWith(string &word){
        Node* node = root;
        for(int i = 0;i<word.size();i++) {
            if(node->containsKey(word[i])){
                node = node->get(word[i]);
            } else {
                return 0;
            }
        }
        return node->getPrefixCount();
    }

    void erase(string &word){
        Node* node = root;
        for(int i = 0;i<word.size();i++) {
            if(node->containsKey(word[i])) {
                node = node->get(word[i]);
                node->decPrefixCount();
            } else {
                return;
            }
        }
        node->decEndsWith();
    }
};

File: Trie/Implement Trie II/Implement_Trie_II.java
// Problem Name: Implement II
// Problem Link: https://www.naukri.com/code360/problems/implement-trie_1387095

import java.util.* ;
import java.io.*;
class Node {
    public Node[] links = new Node[26];
    public int cntEndsWith = 0;
    public int cntPrefix = 0;

    public boolean containsKey(Character ch) {
        return links[ch - 'a'] != null;
    }
    public Node get(Character ch) {
        return links[ch - 'a'];
    }
    public void put(Character ch, Node node) {
        links[ch - 'a'] = node;
    }

    public void incPrefixCount() {
        cntPrefix++;
    }

    public void decPrefixCount() {
        cntPrefix--;
    }

    public void incEndsWithCount() {
        cntEndsWith++;
    }

    public void decEndsWithCount() {
        cntEndsWith--;
    }

    public int getPrefixCount() {
        return cntPrefix;
    }

    public int getEndsWitCount() {
        return cntEndsWith;
    }
}
public class Trie {
    private Node root;
    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        Node node = root;
        for(int i = 0;i<word.length();i++) {
            if(!node.containsKey(word.charAt(i))) {
                node.put(word.charAt(i), new Node());
            }
            node = node.get(word.charAt(i));
            node.incPrefixCount();
        }
        node.incEndsWithCount();
    }

    public int countWordsEqualTo(String word) {
        Node node = root;
        for(int i = 0;i<word.length();i++) {
            if(node.containsKey(word.charAt(i))) {
                node = node.get(word.charAt(i));
            } else {
                return 0;
            }
        }
        return node.getEndsWitCount();
    }

    public int countWordsStartingWith(String word) {
        Node node = root;
        for(int i = 0;i<word.length();i++) {
            if(!node.containsKey(word.charAt(i))) {
                return 0;
            }
            node = node.get(word.charAt(i));
        }
        return node.getPrefixCount();
    }

    public void erase(String word) {
        Node node = root;
        for(int i = 0;i<word.length();i++) {
            if(!node.containsKey(word.charAt(i))) {
                return;
            }
            node = node.get(word.charAt(i));
            node.decPrefixCount();
        }
        node.decEndsWithCount();
    }

}
